File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/tableau-discover.rb
#!/usr/bin/env ruby
# Phase-1 discovery via the Tableau REST API. Use this when you have a PAT —
# it is the FAST path (measured on "Orders Conversion Test": 61.8s serial →
# 13.7–18.9s with the unified pool). The Tableau MCP is the no-PAT fallback.
#
# Output layout (matches what the MCP-driven Phase 1 produces):
#   /tmp/<name>/get-workbook.json     — workbook metadata + view list
#   /tmp/<name>/ds-metadata.json      — VDS read-metadata response (field list + formulas)
#   /tmp/<name>/graphql-fields.json   — metadata API field list (cleaner formulas)
#   /tmp/<name>/views/<viewId>.csv    — every view's data CSV
#   /tmp/<name>/views/<viewId>.png    — dashboard view image only (skip other views by default)
#   /tmp/<name>/workbook-content.twb  — raw .twb XML (or .twbx zip bytes)
#   /tmp/<name>/timings.json          — per-task start/duration/attempts (ALWAYS
#                                       written — the evidence trail for any
#                                       future "discovery is slow" report)
#
# How it fetches: ONE shared thread pool (default 5, --pool N) covers EVERY
# network task — .twb download, VDS read-metadata, GraphQL fields, all view
# CSVs, and the dashboard PNG. Tasks are enqueued longest-job-first (PNG →
# CSVs → twb/VDS/GraphQL) so the slow PNG render hides behind the CSV batch.
# Only the initial workbook GET is serial (everything else needs its view
# list). 5 is the measured sweet spot; 8+ risks long-tail stragglers (a
# contended VizQL session can park one fetch for 40s+).
#
# Resilience (insurance — none of it fired in validation, keep it anyway):
#   * 429 / 408 / 5xx / timeouts retry with exponential backoff + jitter
#     (max 4 attempts).
#   * 401s are re-minted single-flight by lib/tableau_rest.rb (refresh_token!);
#     if a 401 still escapes that retry, the task wrapper re-mints once more.
#
# Usage:
#   eval "$(scripts/get-tableau-token.sh)"
#   ruby scripts/tableau-discover.rb \
#     --workbook-name "Orders Conversion Test" \
#     --datasource-name "ORDER_FACT (CSA.ORDER_FACT)+ (New Virtual Connection)" \
#     --out /tmp/orders [--pool 5]
#
# At least one of --workbook-id / --workbook-name is required.
# --datasource-luid / --datasource-name are optional (auto-detected from the
# .twb when omitted). --datasource-luid must be the FULL UUID — the REST
# filter has no prefix matching.

require 'json'
require 'fileutils'
require 'optparse'
require 'thread'

$LOAD_PATH.unshift File.expand_path('lib', __dir__)
require 'tableau_rest'

opts = { fetch_view_images: 'dashboard-only', pool: 5 }
OptionParser.new do |o|
  o.on('--workbook-name NAME')    { |v| opts[:workbook_name] = v }
  o.on('--workbook-id ID')        { |v| opts[:workbook_id] = v }
  o.on('--datasource-name NAME')  { |v| opts[:datasource_name] = v }
  o.on('--datasource-luid LUID', 'FULL datasource UUID (no prefix matching)') { |v| opts[:datasource_luid] = v }
  o.on('--no-auto-ds', 'Disable .twb-based datasource auto-detect') { opts[:no_auto_ds] = true }
  o.on('--out DIR', 'Output directory (required)') { |v| opts[:out] = v }
  o.on('--skip-images')           { opts[:fetch_view_images] = 'none' }
  o.on('--all-view-images')       { opts[:fetch_view_images] = 'all' }
  o.on('--skip-content')          { opts[:skip_content] = true }
  o.on('--pool N', Integer, 'Fetch-pool size (default 5 — measured sweet spot)') { |v| opts[:pool] = v }
end.parse!

abort 'Missing --out' unless opts[:out]
abort 'Need --workbook-id or --workbook-name' unless opts[:workbook_id] || opts[:workbook_name]

FileUtils.mkdir_p(opts[:out])
FileUtils.mkdir_p(File.join(opts[:out], 'views'))

T0 = Time.now.to_f
TIMINGS = []
TIMINGS_MUTEX = Mutex.new
LOG_MUTEX = Mutex.new

def log(msg)
  LOG_MUTEX.synchronize { warn format('[%7.3f] %s', Time.now.to_f - T0, msg) }
end

# Atomic write: downstream interleaved callers (migrate-tableau.rb) poll for
# these artifacts while this script is still running — never let them observe
# a half-written file.
def atomic_write(path, bytes)
  tmp = "#{path}.tmp.#{Process.pid}"
  File.binwrite(tmp, bytes)
  File.rename(tmp, path)
end

# 400 is included: Tableau Cloud's VizQL layer intermittently 400s view/data
# exports under concurrent load (observed on the FATSCALE 44-view fat workbook:
# 5/44 CSVs 400'd on one run, all succeeded on the next) — a retry with backoff
# recovers them. A genuinely-bad request just burns the 4 attempts.
RETRYABLE = /\b(429|408|400|50[234])\b|Too Many Requests|timed? ?out|Timeout/i

# Run one named fetch task with timing + backoff-retry. Returns task result or nil.
def run_task(name)
  t0 = Time.now.to_f
  attempts = 0
  begin
    attempts += 1
    result = yield
    TIMINGS_MUTEX.synchronize do
      TIMINGS << { 'task' => name, 'start' => (t0 - T0).round(3),
                   'seconds' => (Time.now.to_f - t0).round(3), 'attempts' => attempts, 'ok' => true }
    end
    result
  rescue Tableau::Error, Timeout::Error, Errno::ETIMEDOUT, Net::ReadTimeout, Net::OpenTimeout => e
    msg = e.message.lines.first&.chomp || e.class.name
    if attempts < 4 && msg =~ RETRYABLE
      delay = (1.5 * (2**(attempts - 1))) + rand * 0.5
      log "#{name}: retryable (#{msg[0, 80]}) — backoff #{delay.round(1)}s (attempt #{attempts})"
      sleep delay
      retry
    elsif attempts < 3 && msg =~ /\b401\b/
      # lib already retried 401 once with a refreshed token; one more explicit
      # re-mint covers session invalidation racing across threads.
      log "#{name}: 401 escaped lib retry — re-minting token (attempt #{attempts})"
      (Tableau.refresh_token! rescue nil)
      sleep 1.0 + rand * 0.5
      retry
    end
    TIMINGS_MUTEX.synchronize do
      TIMINGS << { 'task' => name, 'start' => (t0 - T0).round(3),
                   'seconds' => (Time.now.to_f - t0).round(3), 'attempts' => attempts,
                   'ok' => false, 'error' => msg[0, 200] }
    end
    log "#{name}: FAILED after #{attempts} attempt(s): #{msg[0, 120]}"
    nil
  end
end

# --- 1. Workbook (serial — everything else depends on the view list) --------
wb = run_task('get-workbook') do
  w = if opts[:workbook_id]
        Tableau.get_workbook(opts[:workbook_id])
      else
        hit = Tableau.find_workbook_by_name(opts[:workbook_name])
        abort "No workbook found with name=#{opts[:workbook_name]}" unless hit
        Tableau.get_workbook(hit['id'])
      end
  atomic_write(File.join(opts[:out], 'get-workbook.json'), JSON.pretty_generate(w))
  w
end
abort 'workbook fetch failed' unless wb
views = wb.dig('views', 'view') || []
views = [views] unless views.is_a?(Array)
log "wrote get-workbook.json  (id=#{wb['id']} views=#{views.size})"

# --- 1b. Capability banner ---------------------------------------------------
# Tableau Server can lack VDS and the Metadata API that Cloud always exposes.
# Announce the mode up front so an operator isn't left debugging why
# ds-metadata.json / graphql-fields.json never appeared — the .twb XML path
# covers both calc-formula sources when those services are off.
caps = (Tableau.capabilities rescue {})
if caps.any?
  meta_state = caps['metadata_api'] ? 'available' : 'OFF → calc formulas from .twb XML'
  log "capabilities: product=#{caps['product_version'] || '?'} " \
      "REST API=#{caps['rest_api_version'] || '?'}  Metadata API: #{meta_state}"
  log 'note: VDS is probed per-datasource below; on failure discovery falls back to the .twb XML.' unless caps['metadata_api']
end

# --- 2. Build the task queue -------------------------------------------------
queue = Queue.new
twb_done = Queue.new # signals twb completion (for auto-ds fallback)

# 2a. twb download task (.twbx auto-extract preserved from the serial version)
unless opts[:skip_content]
  queue << lambda do
    bytes = run_task('twb-download') { Tableau.download_workbook_content(wb['id']) }
    twb_xml = nil
    if bytes
      if bytes.start_with?("PK\x03\x04")
        twbx_path = File.join(opts[:out], 'workbook-content.twbx')
        atomic_write(twbx_path, bytes)
        log "wrote workbook-content.twbx  (#{bytes.bytesize} bytes)"
        require 'tmpdir'
        Dir.mktmpdir do |tmp|
          unless system('unzip', '-o', '-q', twbx_path, '-d', tmp)
            log '.twbx auto-unzip failed (unzip command not available?); leaving .twbx in place'
          else
            inner = Dir.glob(File.join(tmp, '**', '*.twb')).first
            if inner
              twb_path = File.join(opts[:out], 'workbook-content.twb')
              atomic_write(twb_path, File.binread(inner))
              log "extracted workbook-content.twb  (#{File.size(twb_path)} bytes) from .twbx"
              # Force UTF-8: a bare File.read uses the locale default (US-ASCII on
              # many systems) and raises "invalid byte sequence in US-ASCII" on the
              # first non-ASCII byte (em-dash, @handles, curly quotes). The direct-
              # download branch below already force_encodes; the .twbx path must too.
              twb_xml = File.read(twb_path, encoding: 'UTF-8')
            else
              log '.twbx contained no inner .twb — odd'
            end
          end
        end
      else
        twb_path = File.join(opts[:out], 'workbook-content.twb')
        atomic_write(twb_path, bytes)
        log "wrote workbook-content.twb  (#{bytes.bytesize} bytes)"
        twb_xml = bytes.force_encoding('UTF-8')
      end
    end
    twb_done << twb_xml
  end
end

# 2b. datasource metadata tasks (VDS + GraphQL) — enqueued immediately when a
#     luid or name is supplied; otherwise chained after the twb task (the
#     .twb-caption auto-detect needs the downloaded XML).
ds_metadata_tasks = lambda do |ds_luid|
  queue << lambda do
    vds = run_task('vds-read-metadata') { Tableau.read_metadata(ds_luid) }
    if vds
      atomic_write(File.join(opts[:out], 'ds-metadata.json'), JSON.pretty_generate(vds))
      log "wrote ds-metadata.json  (#{vds.dig('data')&.size || 0} fields)"
    end
  end
  queue << lambda do
    gql = run_task('graphql-fields') { Tableau.graphql_datasource_fields(ds_luid) }
    if gql
      atomic_write(File.join(opts[:out], 'graphql-fields.json'), JSON.pretty_generate(gql))
      log 'wrote graphql-fields.json'
    end
  end
end

ds_luid = opts[:datasource_luid]
if ds_luid.nil? && opts[:datasource_name]
  hit = run_task('find-datasource') { Tableau.find_datasource_by_name(opts[:datasource_name]) }
  ds_luid = hit && hit['id']
end

auto_ds_pending = ds_luid.nil? && !opts[:no_auto_ds] && !opts[:skip_content]
ds_metadata_tasks.call(ds_luid) if ds_luid
if ds_luid.nil? && !auto_ds_pending
  warn 'no --datasource-luid/--datasource-name supplied (and auto-detect is unavailable); skipping VDS + GraphQL fetches'
end

# 2c. dashboard PNG task — IN THE POOL, queued BEFORE the CSVs (longest-job-
#     first: the PNG render is the longest single fetch; starting it at t≈0
#     hides it entirely behind the CSV batch).
case opts[:fetch_view_images]
when 'none'
  log 'skipping view images (--skip-images)'
when 'dashboard-only'
  # Heuristic: the view whose name matches "overview"/"dashboard", else the longest name.
  dash = views.find { |v| v['name'] =~ /\boverview\b|\bdashboard\b/i } ||
         views.max_by { |v| (v['name'] || '').length }
  if dash
    queue << lambda do
      png = run_task("png:#{dash['name']}") { Tableau.view_image(dash['id']) }
      if png
        atomic_write(File.join(opts[:out], 'views', "#{dash['id']}.png"), png)
        log "wrote views/#{dash['id']}.png  (dashboard: #{dash['name']}, #{png.bytesize} bytes)"
      end
    end
  end
when 'all'
  views.each do |v|
    queue << lambda do
      png = run_task("png:#{v['name']}") { Tableau.view_image(v['id']) }
      if png
        atomic_write(File.join(opts[:out], 'views', "#{v['id']}.png"), png)
        log "wrote views/#{v['id']}.png  (#{v['name']})"
      end
    end
  end
end

# 2d. view CSV tasks
views.each do |v|
  queue << lambda do
    csv = run_task("csv:#{v['name']}") { Tableau.view_data(v['id']) }
    if csv
      atomic_write(File.join(opts[:out], 'views', "#{v['id']}.csv"), csv)
      log "wrote views/#{v['id']}.csv  (#{v['name']}, #{csv.bytesize} bytes)"
    end
  end
end

# --- 3. Run the pool ----------------------------------------------------------
n_threads = [opts[:pool], queue.size + 1].min
n_threads = 1 if n_threads < 1
log "pool: #{n_threads} threads, #{queue.size} queued tasks#{auto_ds_pending ? ' (+VDS/GraphQL after twb auto-detect)' : ''}"

# Auto-ds chain: a watcher enqueues VDS/GraphQL once the twb lands.
watcher = nil
unless opts[:skip_content]
  watcher = Thread.new do
    twb_xml = twb_done.pop
    next unless auto_ds_pending
    unless twb_xml
      log 'auto-detect skipped — no .twb content; skipping VDS/GraphQL'
      next
    end
    caption = twb_xml.scan(/<datasource\s+caption='([^']+)'/).flatten
                     .reject { |c| c == 'Parameters' }
                     .first
    if caption
      bare = caption.sub(/\s*\+?\s*\(New Virtual Connection\)\s*$/i, '').strip
      found = nil
      %W[#{caption} #{bare}].uniq.each do |cand|
        hit = run_task("find-datasource:#{cand}") { Tableau.find_datasource_by_name(cand) }
        if hit
          found = hit['id']
          log "auto-detected datasource from .twb: #{cand.inspect} (luid=#{found})"
          break
        end
      end
      ds_metadata_tasks.call(found) if found
      log "could not resolve auto-detected datasource caption #{caption.inspect}; pass --datasource-luid to override" unless found
    else
      log 'auto-detect found no datasource caption in the .twb — skipping VDS/GraphQL'
    end
  end
end

pool = Array.new(n_threads) do
  Thread.new do
    loop do
      task = begin
               queue.pop(true)
             rescue ThreadError
               # queue momentarily empty — the auto-ds watcher may still add
               # tasks; spin until it's dead AND the queue is empty.
               if watcher && watcher.alive?
                 sleep 0.1
                 next
               end
               break
             end
      task.call
    end
  end
end
watcher&.join
pool.each(&:join)

# timings.json is ALWAYS written — it's the evidence trail when someone reports
# discovery slowness later (per-task start offsets show pool occupancy; attempts
# shows whether backoff/re-mint ever fired).
total = (Time.now.to_f - T0).round(3)
atomic_write(File.join(opts[:out], 'timings.json'),
             JSON.pretty_generate('total_seconds' => total, 'pool' => opts[:pool],
                                  'tasks' => TIMINGS.sort_by { |t| t['start'] }))
log "done. total=#{total}s (timings.json written)"

File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/tableau-discover.py
#!/usr/bin/env python3
"""Discover a Tableau workbook through REST without requiring Ruby.

The output contract matches the existing discovery lane: workbook metadata,
thin workbook XML, view CSVs, dashboard PNGs, optional datasource metadata,
and a timings ledger.
"""

from __future__ import annotations

import argparse
import concurrent.futures
import html
import json
import os
import re
import sys
import threading
import time
import zipfile
from pathlib import Path

HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE / "lib"))
import tableau_rest as tableau  # noqa: E402

WRITE_LOCK = threading.Lock()


def as_list(value):
    if value is None:
        return []
    return value if isinstance(value, list) else [value]


def atomic_write(path: Path, content: bytes | str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temp = path.with_name(f"{path.name}.tmp.{os.getpid()}")
    if isinstance(content, bytes):
        temp.write_bytes(content)
    else:
        temp.write_text(content, encoding="utf-8")
    temp.replace(path)


def safe_name(value: str) -> str:
    name = re.sub(r"[^A-Za-z0-9._-]+", "-", value.strip()).strip("-")
    return name[:120] or "dashboard"


def workbook_xml(payload: bytes, out: Path) -> tuple[str, bool]:
    if payload.startswith(b"PK\x03\x04"):
        twbx = out / "workbook-content.twbx"
        atomic_write(twbx, payload)
        with zipfile.ZipFile(twbx) as archive:
            names = archive.namelist()
            inner = next((name for name in names if name.lower().endswith(".twb")), None)
            if not inner:
                raise ValueError("downloaded .twbx contains no .twb workbook")
            xml = archive.read(inner).decode("utf-8-sig", errors="replace")
            has_extract = any(name.lower().endswith(".hyper") for name in names)
    else:
        xml = payload.decode("utf-8-sig", errors="replace")
        has_extract = False
    atomic_write(out / "workbook-content.twb", xml)
    return xml, has_extract


def twb_names(xml: str) -> tuple[list[str], dict[str, list[str]]]:
    worksheets = [
        html.unescape(single or double)
        for single, double in re.findall(
            r"<worksheet\s[^>]*?name=(?:'([^']*)'|\"([^\"]*)\")", xml
        )
    ]
    dashboards: dict[str, list[str]] = {}
    block = re.search(r"<dashboards>.*?</dashboards>", xml, flags=re.DOTALL)
    if not block:
        return worksheets, dashboards
    for segment in re.split(r"(?=<dashboard[\s>])", block.group(0)):
        match = re.match(
            r"<dashboard\s[^>]*?name=(?:'([^']*)'|\"([^\"]*)\")", segment
        )
        if not match:
            continue
        name = html.unescape(match.group(1) or match.group(2))
        zone_names = [
            html.unescape(single or double)
            for single, double in re.findall(
                r"<zone\s[^>]*?name=(?:'([^']*)'|\"([^\"]*)\")", segment
            )
        ]
        dashboards[name] = [item for item in zone_names if item in worksheets]
    return worksheets, dashboards


def match_name(target: str, names: list[str]) -> str | None:
    stripped = target.strip()
    exact = next((name for name in names if name.strip() == stripped), None)
    if exact:
        return exact
    folded = next(
        (name for name in names if name.strip().casefold() == stripped.casefold()),
        None,
    )
    if folded:
        return folded
    partial = [name for name in names if stripped.casefold() in name.casefold()]
    return partial[0] if len(partial) == 1 else None


def scoped_views(
    targets: list[str], views: list[dict], membership: dict[str, list[dict]]
) -> list[dict] | None:
    if not targets or not membership:
        return None
    selected: dict[str, dict] = {}
    for target in targets:
        dashboard = match_name(target, list(membership))
        if not dashboard:
            return None
        for sheet in membership[dashboard]:
            view = next(
                (
                    item
                    for item in views
                    if sheet.get("luid") and item.get("id") == sheet.get("luid")
                ),
                None,
            )
            if view is None:
                view = next(
                    (
                        item
                        for item in views
                        if str(item.get("name", "")).strip().casefold()
                        == str(sheet.get("name", "")).strip().casefold()
                    ),
                    None,
                )
            if view and view.get("id"):
                selected[view["id"]] = view
    return list(selected.values()) or None


class Timings:
    def __init__(self):
        self.started = time.monotonic()
        self.rows: list[dict] = []
        self.lock = threading.Lock()

    def run(self, name, function):
        start = time.monotonic()
        try:
            value = function()
        except Exception as exc:
            with self.lock:
                self.rows.append(
                    {
                        "task": name,
                        "start": round(start - self.started, 3),
                        "seconds": round(time.monotonic() - start, 3),
                        "attempts": 1,
                        "ok": False,
                        "error": str(exc)[:200],
                    }
                )
            raise
        with self.lock:
            self.rows.append(
                {
                    "task": name,
                    "start": round(start - self.started, 3),
                    "seconds": round(time.monotonic() - start, 3),
                    "attempts": 1,
                    "ok": True,
                }
            )
        return value


def discover(args: argparse.Namespace) -> dict:
    out = Path(args.out).expanduser().resolve()
    (out / "views").mkdir(parents=True, exist_ok=True)
    timings = Timings()
    tableau.refresh_token()

    def resolve_workbook():
        if args.workbook_id:
            return tableau.get_workbook(args.workbook_id)
        hit = tableau.find_workbook_by_name(args.workbook_name)
        if not hit:
            raise ValueError(f"no workbook found with name={args.workbook_name!r}")
        return tableau.get_workbook(hit["id"])

    workbook = timings.run("get-workbook", resolve_workbook)
    atomic_write(
        out / "get-workbook.json",
        json.dumps(workbook, indent=2, ensure_ascii=False) + "\n",
    )
    views = as_list((workbook.get("views") or {}).get("view"))

    xml = ""
    has_extract = False
    dashboards: dict[str, list[str]] = {}
    if not args.skip_content:
        payload = timings.run(
            "twb-download",
            lambda: tableau.download_workbook_content(workbook["id"]),
        )
        xml, has_extract = workbook_xml(payload, out)
        _, dashboards = twb_names(xml)

    membership: dict[str, list[dict]] = {}
    if args.dashboard:
        try:
            rows = timings.run(
                "dashboard-membership",
                lambda: tableau.graphql_workbook_dashboards(workbook["id"]),
            )
            membership = {
                str(row.get("name", "")): as_list(row.get("sheets"))
                for row in rows or []
            }
        except tableau.TableauError:
            membership = {
                name: [{"name": sheet, "luid": None} for sheet in sheets]
                for name, sheets in dashboards.items()
            }
    csv_views = scoped_views(args.dashboard, views, membership) or views

    def fetch_csv(view):
        view_id = view["id"]
        body = timings.run(
            f"view-csv:{view_id}", lambda: tableau.view_data(view_id)
        )
        atomic_write(out / "views" / f"{view_id}.csv", body)

    with concurrent.futures.ThreadPoolExecutor(max_workers=args.pool) as pool:
        futures = [pool.submit(fetch_csv, view) for view in csv_views if view.get("id")]
        for future in futures:
            future.result()

    fetched_dashboards = 0
    if not args.skip_images:
        (out / "dashboards").mkdir(parents=True, exist_ok=True)
        dashboard_names = list(dashboards)
        if not dashboard_names:
            dashboard_names = [str(view.get("name", "")) for view in views]
        for dashboard_name in dashboard_names:
            view = next(
                (
                    item
                    for item in views
                    if str(item.get("name", "")).strip().casefold()
                    == dashboard_name.strip().casefold()
                ),
                None,
            )
            if not view or not view.get("id"):
                continue
            image = timings.run(
                f"dashboard-png:{view['id']}",
                lambda view_id=view["id"]: tableau.view_image(view_id),
            )
            atomic_write(
                out / "dashboards" / f"{safe_name(dashboard_name)}.png", image
            )
            atomic_write(out / "views" / f"{view['id']}.png", image)
            fetched_dashboards += 1

    if args.datasource_luid or args.datasource_name:
        datasource = None
        if args.datasource_luid:
            datasource = {"id": args.datasource_luid}
        elif args.datasource_name:
            datasource = tableau.find_datasource_by_name(args.datasource_name)
        if datasource and datasource.get("id"):
            luid = datasource["id"]
            try:
                metadata = timings.run(
                    "datasource-metadata", lambda: tableau.read_metadata(luid)
                )
                atomic_write(
                    out / "ds-metadata.json",
                    json.dumps(metadata, indent=2, ensure_ascii=False) + "\n",
                )
            except tableau.TableauError:
                pass
            try:
                fields = timings.run(
                    "graphql-fields",
                    lambda: tableau.graphql_datasource_fields(luid),
                )
                atomic_write(
                    out / "graphql-fields.json",
                    json.dumps(fields, indent=2, ensure_ascii=False) + "\n",
                )
            except tableau.TableauError:
                pass

    atomic_write(
        out / "timings.json",
        json.dumps(timings.rows, indent=2, ensure_ascii=False) + "\n",
    )
    result = {
        "workbook_id": workbook.get("id"),
        "workbook_name": workbook.get("name"),
        "views": len(views),
        "csv_views": len(csv_views),
        "dashboards_expected": len(dashboards),
        "dashboards_fetched": fetched_dashboards,
        "has_extract": has_extract,
        "workdir": str(out),
    }
    atomic_write(
        out / "discovery-result.json",
        json.dumps(result, indent=2, ensure_ascii=False) + "\n",
    )
    return result


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    source = parser.add_mutually_exclusive_group(required=True)
    source.add_argument("--workbook-name")
    source.add_argument("--workbook-id")
    parser.add_argument("--dashboard", action="append", default=[])
    parser.add_argument("--datasource-name")
    parser.add_argument("--datasource-luid")
    parser.add_argument("--out", required=True)
    parser.add_argument("--pool", type=int, default=5)
    parser.add_argument("--skip-images", action="store_true")
    parser.add_argument("--skip-content", action="store_true")
    args = parser.parse_args()
    if args.pool < 1:
        parser.error("--pool must be at least 1")
    try:
        result = discover(args)
    except (OSError, ValueError, tableau.TableauError) as exc:
        print(f"FATAL: Tableau discovery failed: {exc}", file=sys.stderr)
        return 1
    print(json.dumps(result, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
